sdk/program/src/vote/state/mod.rs

```diff
@@ -55,7 +63,7 @@ impl Vote {
 #[derive(Serialize, Default, Deserialize, Debug, PartialEq, Eq, Copy, Clone, AbiExample)]
 pub struct Lockout {
     pub slot: Slot,
-    pub confirmation_count: u32,
+    confirmation_count: u32,
```

Comment:
I say just make latency a new field and change confirmation count to a u8 like it should have been. @carllin thoughts?

Reply:
> I say just make latency a new field and change confirmation count to a u8 like it should have been. @carllin thoughts?

I'd be happy to, but this is a frozen ABI structure and the data is written into Vote accounts already.  It might be possible to "hack" it though?  With a u8 latency, followed by u16 "for_future_use", followed by a u8 confirmation_count?  Does this sequence:

u8
u16
u8

serialize into the same data as:
u32

as long as the middle bytes of the u32 are always zero?

Reply:
> > how does this work with #27565 ? there are no top 8 bits there
> 
> as is this doesn't work because serialization will fail here https://github.com/solana-labs/solana/blob/master/sdk/program/src/vote/state/mod.rs#L615

That line you reference has been modified in the change and it does work because it calls the confirmation_count() function which returns the confirmation count without the latency embedded in it.  In fact confirmation_count was made into a private struct member so it can't be referenced directly anymore so code that only wants the confirmation_count value cannot mistakenly get a value that includes the latency and confirmation_count conjoined together.

Reply:
Got it, just saw that `VoteStateUpdate` txs don't have latency embedded in them, it's only a part of the `VoteState` on chain. And if i'm understanding this correctly you want to avoid adding a new field to the `VoteState` so you stored this in the unused byte of Lockout?

Reply:
> Got it, just saw that `VoteStateUpdate` txs don't have latency embedded in them, it's only a part of the `VoteState` on chain. And if i'm understanding this correctly you want to avoid adding a new field to the `VoteState` so you stored this in the unused byte of Lockout?

Yes that is exactly correct, but I am very open to different less hacky ways to do it if you can think of any way that doesn't violate the frozen ABI.

Reply:
Can't think of any less hacky ways, if we want to do it properly we'd have to feature flag and use `VoteStateVersions` again. I don't have a strong opinion either way, i'll defer to carl.

Reply:
I'd be OK with the VoteStateVersions solution, but it would require I believe moving the latencies out to a separate data structure not within Lockouts and maintaining it in parallel to Lockouts.  I say this because I believe that Lockouts being ABI frozen means it can't change at all, although the VoteState containing it can, which is where the parallel data structure would be placed I guess.

Reply:
Yeah I think we'd change it to be `VecDeque<(Lockout, Latency)>` or something, i'd rather not augment `Lockout` since it also used for txs and latency seems to be a function of the `VoteState` itself.

Reply:
Hi, is there consensus on what this change would need to look like to be accepted?  Is re-using the upper bits of Lockout not acceptable?  I will adjust this change according to your feedback.  Thanks!

Reply:
@carllin wdyt?

Reply:
I would say we should migrate to a new VoteStateVersion for maintainability, hacking the top bits works but seems a bit brittle

Reply:
VoteState is frozen ABI - does that mean it can be changed to a new VoteStateVersion or not?

I don't really know how your frozen ABI stuff works.  I know that there is a V0_23_5 VoteState but I don't know if that split happened before the ABI was frozen, or after.  If it was after, then obviously a new VoteState can be added.  If it was after, then -- does that mean that a new version can't be made since the ABI is frozen?

If it **is** possible to make a new vote state version, what is the point of the 'frozen ABI' of vote state?

Or maybe frozen ABI means: "you can only change this data structure using the version method like was done with VoteStateVersion".

Also - what version number would the current VoteStateVersion get?  The previous old one was VoteState0_23_5.  Would I be taking a snapshot of master's VoteState and calling that ... VoteState1_10_40?  Or VoteState1_13_2?  Or something else?

Reply:
Also since VoteState is already written into vote accounts, how does new space get added for any new members of VoteState?

Is there some kind of upgrade method that happens when the validator is starting up that would update all Vote accounts or sometihng like that?

Reply:
Hm I think it's simpler than I envisioned.  As long as vote accounts have enough space to store the new values, then I think the "upgrade" happens as each vote account is loaded, and the VoteStateVersions logic upgrades the previous versions to the new version; then when the new vote state is written, the new version is written.  The only issue is, do existing vote accounts have room for a new VecDeque<u8>?  I think they do given that all vote accounts on mainline are length 3731, and it appears that they all have a few hundred bytes to spare.

Reply:
OK well I've updated the implementation to use a new VoteStateVersion instead of re-using bits from Lockout.confirmation_count.

I did not know what to do with the VoteState ABI version.  I don't know if it needs to change somehow.  But the build succeeds and tests succeeded, so I guess not?

Reply:
Turns out I had to make a new Tower version also since the tower includes an embedded VoteState.

Reply:
Still working through CI breaks.  I can't reproduce these locally, the setup to get all the tests to run is beyond me.

Reply:
I don't know how to interpret the latest CI build failure.  There is a lot of text there and it's hard for me to discern if the failure has something to do with my change or is an artifact of an unreliable build.

Reply:
@bji I think the best would be to 1st get the proposal discussed and approved then investigate best way to implement that.
I see you opened two PR for the proposal but closed one. Is #28162 the current updated one?

To avoid confusion it would be best if you just update that on PR instead of creating a new PR. thank you,

Reply:
I am looking at the proposal right now.
@AshwinSekar @carllin Can you guys also take a look at the proposal #28162 and comment if needed? thanks

Reply:
> I am looking at the proposal right now. @AshwinSekar @carllin Can you guys also take a look at the proposal #28162 and comment if needed? thanks

I am updating the proposal now because it has a brief section at the end which mentions specific implementation details which are now out of date.  This should not materially impact the proposal, it's just changing some notes about expected implementation.